serverless/cmd/feeder/github/main.go

```diff
@@ -0,0 +1,561 @@
+const usage = `Usage:
+ feed-to-github --log_owner_repo --witness_owner_repo --log_repo_path --feeder_config_file --interval
+
+Where:
+ --log_owner_repo is the repo owner/fragment from the repo URL.
+     e.g. github.com/AlCutter/serverless-test -> AlCutter/serverless-test
+ --witness_owner_repo is the repo owner/fragment of the forked log to use for the PR branch.
+ --log_repo_path is the path from the root of the repo where the log files can be found,
+ --feeder_config_file is the path to the config file for the serverless/cmd/feeder command.
+ --interval if set, the script will continously feed and (if needed) create witness PRs sleeping
+     the specified number of seconds between attempts. If not provided, the tool does a one-shot feed.
+
+`
+
+var (
+	logOwnerRepo     = flag.String("log_owner_repo", "", "The repo owner/fragment from the log repo URL.")
+	witnessOwnerRepo = flag.String("witness_owner_repo", "", "The repo owner/fragment from the witness (forked log) repo URL.")
+	logRepoPath      = flag.String("log_repo_path", "", "Path from the root of the repo where the log files can be found.")
+	feederConfigPath = flag.String("feeder_config_file", "", "Path to the config file for the serverless/cmd/feeder command.")
+	interval         = flag.Duration("interval", time.Duration(0), "Interval between checkpoints.")
+	cloneToDisk      = flag.Bool("clone_to_disk", false, "Whether to clone the log repo to memory or disk.")
+)
+
+func main() {
+	flag.Parse()
+	opts := mustConfigure()
+	ctx := context.Background()
+
+	if *cloneToDisk {
+		// Make a tempdir to clone the witness (forked) log into.
+		tmpDir, err := os.MkdirTemp(os.TempDir(), "feeder-github")
+		if err != nil {
+			glog.Exitf("Error creating temp dir: %v", err)
+		}
+		opts.witnessClonePath = tmpDir
+		defer func() {
+			if err := os.RemoveAll(tmpDir); err != nil {
+				glog.Warningf("RemoveAll err: %v", err)
+			}
+		}()
+	}
+
+	witnessRepo, err := setupWitnessRepo(ctx, opts)
+	if err != nil {
+		glog.Exitf("Error during set-up of witness repo: %v", err)
+	}
+
+	// Authenticate to Github.
+	ghClient, err := authWithGithub(ctx, opts.githubAuthToken)
+	if err != nil {
+		glog.Exitf("Error authenticating to Github: %v", err)
+	}
+	glog.V(1).Info("Created Github client")
+
+	// Overview of the main feeder loop:
+	//   1. run feeder to check if there are new signatures
+	//   2. if not, sleep until next time to check.
+	//   3. if so, move output from feeder onto new branch
+	//   4. make commit to branch
+	//   5. create checkpoint PR
+	//   6. clean up
+	//   7. --end of cycle--
+
+	fmt.Println("[Starting feeding]---------------------------------------------------")
+	for {
+		if err = feedOnce(ctx, opts, witnessRepo, ghClient); err != nil || opts.feedInterval == 0 {
+			break
+		}
+		fmt.Println("[Feed cycle complete]------------------------------------------------")
+
+		<-time.After(opts.feedInterval)
+	}
+	if err != nil {
+		glog.Exitf("CreateCheckpointPR error: %v", err)
```

Comment:
`CreateCheckpointPR` seems like it's probably a layer violation here (probably because a method has now been extracted). `feedOnce` would be easier to debug from looking at this code as it appears right now.

Reply:
yup - done